src/Spanner/Session/CacheSessionPool.php

```diff
@@ -0,0 +1,548 @@
+<?php
+/**
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace Google\Cloud\Spanner\Session;
+
+use Google\Cloud\Core\Lock\LockInterface;
+use Google\Cloud\Core\Lock\SymfonyLockAdapter;
+use Google\Cloud\Spanner\Database;
+use Psr\Cache\CacheItemPoolInterface;
+use Symfony\Component\Lock\Factory;
+use Symfony\Component\Lock\Store\FlockStore;
+
+/**
+ * This session pool implementation accepts a PSR-6 compatible cache
+ * implementation and utilizies it to store sessions between requests.
+ *
+ * Please note that if you configure a high minimum session value the first
+ * request and any after a period of inactivity greater than an hour (the point
+ * at which sessions will expire) will have an increased amount of latency. This
+ * is due to the pool attempting to create as many sessions as needed to fill
+ * itself to match the minimum value.
+ *
+ * For this reason, it is highly recommended to configure a script to make an
+ * initial request to warm up the pool and manage subsequent requests during
+ * off-peak hours to keep the pool active.
+ *
+ * Please note: While required for the session pool, a PSR-6 implementation is
+ * not included in this library. It will be neccesary to include a separate
+ * dependency to fulfill this requirement.
+ *
+ * Example:
+ * ```
+ * use Google\Cloud\Spanner\SpannerClient;
+ * use Google\Cloud\Spanner\Session\CacheSessionPool;
+ * use Symfony\Component\Cache\Adapter\FilesystemAdapter;
+ *
+ * $spanner = new SpannerClient();
+ * $cache = new FilesystemAdapter();
+ * $sessionPool = new CacheSessionPool($cache);
+ *
+ * $database = $spanner->connect('my-instance', 'my-database', [
+ *     'sessionPool' => $sessionPool
+ * ]);
+ * ```
+ */
+class CacheSessionPool implements SessionPoolInterface
+{
+    const CACHE_KEY_TEMPLATE = 'cache-session-pool.%s.%s';
+
+    /**
+     * @var array
+     */
+    private static $defaultConfig = [
+        'maxSessions' => PHP_INT_MAX,
```

Comment:
Does the spanner backend have a limit for number of sessions? If the spanner has the limit, we should have smaller limit than that.

However, if you limit the number of sessions it creates at one time (as discussed below), I'm fine with keeping the `PHP_INT_MAX`.